### DISCUSSION TITLE
How to implement custom hardening stress update material?

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#29096] [Q&A Modules: Solid mechanics]
I am trying to implement a custom hardening stress update material similar to isotropic power law hardening material. But some parts in isotropic power law hardening are quite confusing. 
```
template <bool is_ad>
GenericReal<is_ad>
IsotropicPowerLawHardeningStressUpdateTempl<is_ad>::computeHardeningDerivative(
    const GenericReal<is_ad> & scalar)
{
  const GenericReal<is_ad> stress_delta = _effective_trial_stress - _three_shear_modulus * scalar;
  GenericReal<is_ad> slope = std::pow(stress_delta, (1.0 / _strain_hardening_exponent - 1.0)) /
                             _strain_hardening_exponent * 1.0 /
                             std::pow(_K, 1.0 / _strain_hardening_exponent);
  slope -= 1.0 / _youngs_modulus;

  return 1.0 / slope;
}
```
Here why is the hardening derivative such a complex form instead of n*K*p^n-1? And why is there a reciprocal of slope? I also tried derivative as n*K*p^n-1, but the convergence is bad. Can some expert please explain this?

**Comment by GiudGiud** [comment]
Hello

I think it is like this?
```
std::pow(stress_delta, (1.0 / _strain_hardening_exponent - 1.0)) = p ^ (n - 1)
with n = 1.0 / _strain_hardening_exponent
1 / _strain_hardening_exponent = n as a factor in the expression
 std::pow(_K, 1.0 / _strain_hardening_exponent);  = K
```
Then the young modulus is the elastic part of that slope

>  I also tried derivative as nKp^n-1, but the convergence is bad

The way you can prove to yourself this is correct is to use the Jacobian analysis.(if this slope term is for the Jacobian, which I imagine it is)
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

Guillaume
